### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by TommyM0ntana**

### CODE REVIEW

File: enumerables.rb
Comment (uche-inyama) [inline enumerables.rb:4]: There should be a condition after the line 4, that returns an enumerable, when no block is given

File: enumerables.rb
Comment (uche-inyama) [inline enumerables.rb]: You should an enumerable, if block is not given after defining the method. Also check the spelling of index- in your method name.

File: enumerables.rb
Comment (uche-inyama) [inline enumerables.rb]: my_select method should also return an enumerable if no block is given.
check your code on line 24, it should be **```result.push```**

File: enumerables.rb
Comment (uche-inyama) [inline enumerables.rb]: my_all method has three conditions; when a block is given, when a pattern is given and when no pattern or block is given.

However, use this code snippet as a guide.

```  
def my_all?(pattern = nil)
    result = true
    if block_given?
      my_each { |ele| result &= (yield ele) }
    elsif pattern
      my_each { |ele| result &= pattern === ele }
    else
      my_each { |ele| result &= ele }
    end
    result
  end 
```

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Hi Tommy!

After passing your module through test, your methods failed the test. 

I have left some comments below on how to go about resolving the issue. But importantly, use this [link](https://github.com/igakigongo/EnumerableMethods/blob/feature/enumerable-unit-tests/enumerable.rb)  :100:  as a guide to implement the changes.
- [ ] Also, ensure linter checks are passed.

After implementing the changes submit another form for a code review

Also, if you need clarification you can reach me on slack, and we can schedule a call :slightly_smiling_face: 

#### ``` Have a great day``` :slightly_smiling_face: 

@ Uchechukwu Henry Inyama

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Hi Montana,

You have not implemented the changes correctly,

For example:

- [ ]  On line 30, you did ```return.push``` , instead of ```result.push```

- [ ] On line 69, you did not close your if block.

Take your time to implement the changes. Also, some of the tests are still failing.

Implement the changes and submit another form for a code review

#### ```Have a great day ``` :slightly_smiling_face: 

@ Uchechukwu Henry Inyama

File: enumerables.rb
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):009:0> array.all?(Integer)
=> true
irb(main):010:0> array.my_all?(Integer)
=> false
irb(main):011:0> 
irb(main):012:0> 
irb(main):013:0> array
=> [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]
irb(main):014:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):016:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):017:0> words.all?(/d/)
=> true
irb(main):018:0> words.my_all?(/d/)
=> false
irb(main):019:0>
```

File: enumerables.rb
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):021:0> array
=> [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]
irb(main):022:0> array.any?(Integer)
=> true
irb(main):023:0> array.my_any?(Integer)
=> false
irb(main):024:0>
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):026:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):027:0> words.any?(/d/)
=> true
irb(main):028:0> words.my_any?(/d/)
=> false
irb(main):029:0>
```

File: enumerables.rb
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):031:0> true_block = proc { |num| num > 9 }
=> #<Proc:0x00005608bbe3bd08@(irb):31>
irb(main):033:0> array.none?(&true_block)
=> true
irb(main):034:0> array.my_none?(&true_block)
=> false
irb(main):035:0>
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):037:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):038:0> words.none?(/z/)
=> true
irb(main):039:0> words.my_none?(/z/)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):39
        1: from enumerables.rb:62:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):040:0> 
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):042:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):043:0> words.none?(5)
=> true
irb(main):044:0> words.my_none?(5)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):44
        1: from enumerables.rb:62:in `my_none?'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):045:0>
```

File: enumerables.rb
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):047:0> array
=> [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]
irb(main):048:0> array.count
=> 100
irb(main):049:0> array.my_count
=> 0
irb(main):050:0>
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):052:0> block = proc { |num| num < 4 }
=> #<Proc:0x00005608bbe7d820@(irb):52>
irb(main):053:0> array
=> [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]
irb(main):054:0> array.count(3)
=> 13
irb(main):055:0> array.my_count(3)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):55
        1: from enumerables.rb:69:in `my_count'
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):056:0> 
```

File: enumerables.rb
Comment (maelfosso) [inline enumerables.rb]: Should return an enumerator
```ruby
irb(main):058:0> array
=> [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]
irb(main):059:0> array.map
=> #<Enumerator: [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]:map>
irb(main):060:0> array.my_map
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):60
        3: from enumerables.rb:80:in `my_map'
        2: from enumerables.rb:9:in `my_each'
        1: from enumerables.rb:80:in `block in my_map'
LocalJumpError (no block given (yield))
irb(main):061:0> 
```

File: enumerables.rb
Comment (maelfosso) [inline enumerables.rb]:
```ruby
irb(main):063:0> range = Range.new(5, 50)
=> 5..50
irb(main):064:0> range.my_inject(4) { |prod, n| prod * n }
Traceback (most recent call last):
        4: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        3: from (irb):64
        2: from enumerables.rb:85:in `my_inject'
        1: from enumerables.rb:9:in `my_each'
NoMethodError (undefined method `[]' for 5..50:Range)
irb(main):065:0> range.inject(4) { |prod, n| prod * n }
=> 5069015533618896340602101361010794807396273594826752000000000000
```

Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):069:0> array
=> [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]
irb(main):070:0> array.inject(:+)
=> 384
irb(main):071:0> array.my_inject(:+)
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):71
        3: from enumerables.rb:85:in `my_inject'
        2: from enumerables.rb:9:in `my_each'
        1: from enumerables.rb:85:in `block in my_inject'
LocalJumpError (no block given (yield))
irb(main):072:0> 
```

File: enumerables.rb
Comment (maelfosso) [reply enumerables.rb]:
```ruby
irb(main):075:0> block = proc { |num| num < 4 }
=> #<Proc:0x00005608bbdf3350@(irb):75>
irb(main):076:0> array.count(&block)
=> 49
irb(main):077:0> array.my_count(&block)
=> [4, 3, 2, 6, 2, 8, 8, 3, 0, 4, 8, 3, 4, 8, 1, 5, 2, 5, 6, 3, 1, 8, 2, 1, 1, 5, 3, 6, 0, 4, 8, 6, 7, 6, 2, 4, 3, 8, 1, 3, 5, 1, 0, 2, 6, 1, 1, 2, 2, 1, 7, 5, 4, 5, 2, 3, 5, 1, 3, 4, 1, 4, 3, 2, 0, 5, 4, 7, 1, 4, 1, 1, 8, 6, 8, 7, 3, 7, 6, 3, 6, 8, 4, 1, 1, 3, 5, 8, 7, 1, 2, 2, 4, 5, 7, 2, 5, 8, 0, 0]
irb(main):078:0>
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required

Hello @Tommy

Nice improvements :clap: but there are some changes to do.
Please, read the comments below to know more about it.

Kindly  make the changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: enumerables.rb
Comment (ebukaume) [inline enumerables.rb]: Your module could not run; you have error in your code.

On line 74, are you starting a new if block without closing the preceding one? Or were you trying to write `elsif`?

**Review by ebukaume — CHANGES_REQUESTED** [review]
## Changes Required!

Hi @TommyM0ntana :wave: 

You are doing well implementing those requested changes. However, there is an error in your code that is preventing me from proceeding with the review. **Please refer to my comment below and submit another review request after fixing it.** 

_I'd recommend you always test your codes before submitting for review_

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

File: enumerables.rb
Comment (TommyM0ntana) [reply enumerables.rb]: Can you please tell me how to fix rubocop., The cose if fine but it gives Cyclomatic complexity.About the new code review

Comment (ebukaume) [reply enumerables.rb]: When faced with cyclomatic complexity, you should try to rethink your logic. Most times, there are simpler ways to structure the conditionals.

Comment (TommyM0ntana) [reply enumerables.rb]: i fix it but now it says that the line is too long!!!

File: enumerables.rb
Comment (ebukaume) [inline enumerables.rb]: Remember that `my_all?` is supposed to return either `true` or `false`. However, line 41 returns `nil`.

Also, did you consider when the user passes a proc as an argument?

Finally, where did you define the `result` you are using on lines 43 and 45? It is blowing up the method execution, thus I can't test the method properly.

File: enumerables.rb
Comment (ebukaume) [inline enumerables.rb]: my_proc = proc { |num| num > 100 }
[1, 2, 3].my_any?(&my_proc) returns `true` instead of `false`; please fix it too.

File: enumerables.rb
Comment (ebukaume) [inline enumerables.rb]: You have a typo on line 66

File: enumerables.rb
Comment (ebukaume) [inline enumerables.rb]: What happens when a range is passed? Remember that the Range class also includes the enumerable method.

**Review by ebukaume — CHANGES_REQUESTED** [review]
## Changes required! :stop_sign: 

Hi @TommyM0ntana,

Good job fixing the issue with your conditional. Please see below for more.

- [ ] Also remember to make the linter pass

**Kindly submit another code review request after fixing those issues.** :point_down: 

Happy Coding :v: 
**Ebuka Umeokonkwo**
 [Github](https://github.com/ebukaume) | [Linkedin](https://www.linkedin.com/in/ebukaume) | [Slack](https://app.slack.com/team/UFDKXE68L) | [Twitter](https://twitter.com/ebukaume)

File: enumerables.rb
Comment (ebukaume) [reply enumerables.rb]: If you can not optimize the logic further, consider extracting some parts into a private method.

File: enumerables.rb
Comment (stratospherique) [inline enumerables.rb]: * [x] my_none? returns true if the block never returns true for all elements
  ```ruby
    array = [1,5,4,8,7,9,5,2,0,4]
    true_block = proc { |num| num > 9}
    array.my_none?(&true_block) == array.none?(&true_block)
  ```

File: enumerables.rb
Comment (stratospherique) [inline enumerables.rb:114]: * [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5,50);
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
     
    # NoMethodError:
    #   undefined method `[]' for 5..50:Range
     # ./enumerables.rb:11:in `my_each'
     # ./enumerables.rb:111:in `my_inject'
  ```
     
* [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5,50);
    p range.my_inject(2, :*) # returns this error
     
    # NoMethodError:
    #   undefined method `[]' for 5..50:Range
     # ./enumerables.rb:11:in `my_each'
     # ./enumerables.rb:111:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Impressive Improvement @TommyM0ntana :clap: You are almost there. However, few issues are still to be solved in order to get approval for this project :arrow_down: 

- **[REQUIRED]** Check the code comment below, use the provided examples to refactor your methods :nerd_face: 

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Review by uche-inyama — CHANGES_REQUESTED** [review]
## Change required.

Hi Montana!

Please fix the conflicting file, then submit another form for a code review. I wasn't able to test the method.

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Review by stratospherique — APPROVED** [review]
### Great work implementing the suggested :clap: Congrats, the project is approved :heavy_check_mark: You may merge :+1: :+1: :+1: 

Keep up the good work :muscle: 
**_Ahmed Mahfoudh_**

###### Before you merge, you should know that for this project, you were supposed to follow the git flow guidelines suggested by microverse https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/. To rectify the issue, I suggest that you create a dev-branch and make it as the default branch. Then, change the target branch of this pull request to be **dev-branch** instead of **master**. After doing these steps, you can safely merge this pull request

**Pull request merged**
